org.eclipse.xtext.xtext.ui.examples/projects/fowlerdsl/org.eclipse.xtext.example.fowlerdsl.ui.tests/src/org/eclipse/xtext/example/fowlerdsl/ui/tests/StatemachineFoldingTest.xtend

```diff
@@ -0,0 +1,92 @@
+/*******************************************************************************
+ * Copyright (c) 2019 itemis AG (http://www.itemis.eu) and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *******************************************************************************/
+package org.eclipse.xtext.example.fowlerdsl.ui.tests
+
+import org.eclipse.xtext.testing.InjectWith
+import org.eclipse.xtext.testing.XtextRunner
+import org.eclipse.xtext.ui.testing.AbstractFoldingTest
+import org.junit.Test
+import org.junit.runner.RunWith
+
+/**
+ * @author miklossy - Initial contribution and API
+ */
+@RunWith(XtextRunner)
+@InjectWith(StatemachineUiInjectorProvider)
+class StatemachineFoldingTest extends AbstractFoldingTest {
+
+	@Test def state001() {
+		'''
+			state idle
+			end
+		'''.hasFoldingRegions(1 -> 2)
+	}
+
+	@Test def state002() {
+		'''
+			state idle
+				
+				
+				
+			end
+		'''.hasFoldingRegions(1 -> 5)
+	}
+
+	@Test def complex() {
+		'''
+			events
+				doorClosed   D1CL
+				drawerOpened D2OP
+				lightOn      L1ON
+				doorOpened   D1OP
+				panelClosed  PNCL
+			end
+			
+			resetEvents
+				doorOpened
+				doorClosed
+			end
+			
+			commands
+				unlockPanel PNUL
+				lockPanel   NLK
+				lockDoor    D1LK
+				unlockDoor  D1UL
+			end
+			
+			state idle
+				actions {unlockDoor lockPanel}
+				doorClosed => active
+			end
+			
+			state active
+				drawerOpened => waitingForLight
+				lightOn      => waitingForDrawer
+			end
+			
+			state waitingForLight
+				lightOn => unlockedPanel
+			end
+			
+			state waitingForDrawer
+				drawerOpened => unlockedPanel
+			end
+			
+			state unlockedPanel
+				actions {unlockPanel lockDoor}
+				panelClosed => idle
+			end
+		'''.hasFoldingRegions(
+			21 -> 24,
```

Comment:
I find this very hard to read. Could we annotate the DSL text to mark folding regions?

Reply:
Actually that was also my first idea to annotate the DSL text with the `«c»` special symbols (similar to the other ui test cases) to mark the start end the end positions of the expected folding regions. However, that works only if the expected folding regions are disjunct regions. Considering the corner cases (e.g: overlapping regions, duplicated regions, ...) the annotated DSL would become much more difficult to read.